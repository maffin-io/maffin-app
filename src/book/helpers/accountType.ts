import type { Account } from '@/book/entities';

export function isInvestment(account: Account): boolean {
  if (['STOCK', 'MUTUAL'].includes(account.type)) {
    return true;
  }

  return false;
}

export function isAsset(account: Account): boolean {
  if (['ASSET', 'BANK', 'CASH', 'EQUITY'].includes(account.type)) {
    return true;
  }

  return false;
}

export function getAllowedSubAccounts(type: string): string[] {
  if (type === 'ROOT') {
    return ['ASSET', 'EXPENSE', 'INCOME', 'LIABILITY'];
  }

  if (type === 'ASSET') {
    return ['ASSET', 'BANK', 'CASH', 'EQUITY', 'STOCK', 'MUTUAL'];
  }

  return [type];
}
