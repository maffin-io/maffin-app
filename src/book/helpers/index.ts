export { currencyToSymbol } from './currencyToSymbol';
export { isInvestment, isLiability, isAsset } from './accountType';
