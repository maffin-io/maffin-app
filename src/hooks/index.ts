export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
