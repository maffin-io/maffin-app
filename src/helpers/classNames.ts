import classNames from 'classnames';

import { isAsset, isInvestment, isLiability } from '@/book/helpers/accountType';
import type { Account } from '@/book/entities';

export function accountColorCode(account?: Account, defaults?: string) {
  return classNames(defaults, {
    misc: account && isInvestment(account),
    default: account?.type === 'EQUITY',
    success: account?.type === 'INCOME',
    danger: account?.type === 'EXPENSE',
    warning: account && isLiability(account),
    info: account && isAsset(account),
  });
}
