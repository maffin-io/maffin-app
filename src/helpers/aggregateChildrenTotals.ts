import { DateTime } from 'luxon';

import Money from '@/book/Money';
import { Account, Commodity } from '@/book/entities';
import type { AccountsTotals, AccountsMap } from '@/types/book';
import type { PriceDBMap } from '@/book/prices';
import mapAccounts from './mapAccounts';

/**
 * Given an AccountsTotal, aggregate the totals of the children into
 * their the account identified by 'guid'.
 *
 * If the children have different commodities, convert
 * them accordingly.
 */
export default function aggregateChildrenTotals(
  guid: string,
  accounts: Account[],
  prices: PriceDBMap,
  selectedDate: DateTime,
  totals: AccountsTotals,
): AccountsTotals {
  const accountsMap = mapAccounts(accounts);
  const aggregatedTotals: AccountsTotals = {};
  accountsMap[guid].childrenIds.forEach((childId: string) => {
    aggregate(
      childId,
      accountsMap,
      prices,
      selectedDate,
      totals,
      aggregatedTotals,
    );

    aggregatedTotals[`type_${accountsMap[childId].type.toLowerCase()}`] = aggregatedTotals[childId];
  });

  return aggregatedTotals;
}

function aggregate(
  guid: string,
  accounts: AccountsMap,
  prices: PriceDBMap,
  selectedDate: DateTime,
  totals: AccountsTotals,
  aggregatedTotals: AccountsTotals,
): Money {
  const current = accounts[guid];
  aggregatedTotals[current.guid] = totals[current.guid] || new Money(0, current.commodity.mnemonic);

  current.childrenIds.forEach((childId: string) => {
    aggregatedTotals[current.guid] = aggregatedTotals[current.guid].add(
      convertToParentCommodity(
        aggregate(childId, accounts, prices, selectedDate, totals, aggregatedTotals),
        accounts[childId].commodity,
        current.commodity,
        prices,
        selectedDate,
      ),
    );
  });

  return aggregatedTotals[current.guid];
}

/**
 * Given a Money amount it converts it to the specified currency in parameter
 * to.
 *
 * If the passed commodity in from is an investment, we retrieve the price of the
 * stock and if the currency of the stock doesn't match the specified currency,
 * we re-convert again
 */
function convertToParentCommodity(
  amount: Money,
  from: Commodity,
  to: Commodity,
  prices: PriceDBMap,
  selectedDate: DateTime,
): Money {
  let rate = 1;
  let currency = from;

  if (currency.namespace !== 'CURRENCY') {
    const stockPrice = prices.getInvestmentPrice(
      currency.mnemonic,
      selectedDate,
    );
    rate = stockPrice.value;
    currency = stockPrice.currency;
  }
  if (currency.guid !== to.guid) {
    rate *= prices.getPrice(
      currency.mnemonic,
      to.mnemonic,
      selectedDate,
    ).value;
  }

  return amount.convert(to.mnemonic, rate);
}
