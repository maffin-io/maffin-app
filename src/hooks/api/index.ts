import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
