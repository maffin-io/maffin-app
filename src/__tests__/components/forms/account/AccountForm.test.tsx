import React from 'react';
import { DateTime } from 'luxon';
import {
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DataSource } from 'typeorm';
import * as navigation from 'next/navigation';
import type { AppRouterInstance } from 'next/dist/shared/lib/app-router-context.shared-runtime';
import { UseQueryResult } from '@tanstack/react-query';

import { getAllowedSubAccounts } from '@/book/helpers/accountType';
import {
  Account,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import AccountForm from '@/components/forms/account/AccountForm';
import * as apiHook from '@/hooks/api';

jest.mock('next/navigation');
jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

describe('AccountForm', () => {
  let datasource: DataSource;
  let eur: Commodity;
  let root: Account;
  let assetAccount: Account;
  let expenseAccount: Account;

  beforeEach(async () => {
    jest.spyOn(DateTime, 'now').mockReturnValue(DateTime.fromISO('2023-01-30') as DateTime<true>);
    datasource = new DataSource({
      type: 'sqljs',
      dropSchema: true,
      entities: [Account, Commodity, Price, Split, Transaction],
      synchronize: true,
      logging: false,
    });
    await datasource.initialize();

    eur = await Commodity.create({
      namespace: 'CURRENCY',
      mnemonic: 'EUR',
    }).save();

    root = await Account.create({
      name: 'Root',
      type: 'ROOT',
    }).save();

    assetAccount = await Account.create({
      name: 'Assets',
      type: 'ASSET',
      fk_commodity: eur,
      parent: root,
    }).save();

    expenseAccount = await Account.create({
      name: 'Expenses',
      type: 'EXPENSE',
      fk_commodity: eur,
      parent: root,
    }).save();

    root.path = 'Root';
    assetAccount.path = 'Assets';
    expenseAccount.path = 'Expenses';

    jest.spyOn(apiHook, 'useCommodities').mockReturnValue({ data: [eur] } as UseQueryResult<Commodity[]>);
    jest.spyOn(apiHook, 'useAccounts').mockReturnValue({
      data: [root, assetAccount, expenseAccount],
    } as UseQueryResult<Account[]>);
    jest.spyOn(apiHook, 'useMainCurrency')
      .mockReturnValue({ data: undefined } as UseQueryResult<Commodity>);
  });

  afterEach(async () => {
    jest.resetAllMocks();
    await datasource.destroy();
  });

  it('renders as expected with add', async () => {
    const { container } = render(
      <AccountForm
        action="add"
        onSave={() => {}}
      />,
    );

    await screen.findByLabelText('Name');
    screen.getByRole('combobox', { name: 'parentInput' });
    screen.getByRole('combobox', { name: 'commodityInput' });
    expect(screen.getByLabelText('typeInput')).toBeDisabled();
    screen.getByRole('spinbutton', { name: 'Opening balance', hidden: true });
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with update', async () => {
    const { container } = render(
      <AccountForm
        action="update"
        onSave={() => {}}
      />,
    );

    await screen.findByLabelText('Name');
    screen.getByRole('combobox', { name: 'parentInput' });
    expect(screen.getByLabelText('typeInput')).toBeDisabled();
    expect(screen.getByLabelText('commodityInput')).toBeDisabled();
    screen.getByRole('spinbutton', { name: 'Opening balance', hidden: true });
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with delete', async () => {
    const { container } = render(
      <AccountForm
        action="delete"
        onSave={() => {}}
      />,
    );

    await screen.findByLabelText('Name');
    expect(screen.getByLabelText('parentInput')).toBeDisabled();
    expect(screen.getByLabelText('typeInput')).toBeDisabled();
    expect(screen.getByLabelText('commodityInput')).toBeDisabled();
    screen.getByRole('spinbutton', { name: 'Opening balance', hidden: true });
    expect(container).toMatchSnapshot();
  });

  it('sets mainCurrency as default commodity', async () => {
    jest.spyOn(apiHook, 'useMainCurrency')
      .mockReturnValue({ data: eur } as UseQueryResult<Commodity>);
    render(
      <AccountForm
        action="add"
        onSave={() => {}}
      />,
    );

    await screen.findByText('EUR');
  });

  it('overrides mainCurrency with passed default', async () => {
    jest.spyOn(apiHook, 'useMainCurrency')
      .mockReturnValue({ data: eur } as UseQueryResult<Commodity>);
    render(
      <AccountForm
        action="add"
        defaultValues={{
          fk_commodity: {
            mnemonic: 'USD',
          } as Commodity,
        }}
        onSave={() => {}}
      />,
    );

    await screen.findByText('USD');
  });

  it.each([
    'INCOME', 'EXPENSE',
  ])('hides commodity field when %s', async (type) => {
    jest.spyOn(apiHook, 'useMainCurrency')
      .mockReturnValue({ data: eur } as UseQueryResult<Commodity>);
    render(
      <AccountForm
        action="add"
        defaultValues={{
          type,
        }}
        onSave={() => {}}
      />,
    );

    const fieldsets = screen.getAllByRole('group');
    expect(fieldsets[5]).toHaveClass('hidden');
  });

  it('renders with defaults as expected', async () => {
    render(
      <AccountForm
        defaultValues={{
          name: 'Test account',
          parent: assetAccount,
          type: 'BANK',
          fk_commodity: eur,
        }}
        onSave={() => {}}
      />,
    );

    await waitFor(() => expect(screen.getByLabelText('Name')).toHaveValue('Test account'));
    screen.getByText('Assets');
    screen.getByText('BANK');
    screen.getByText('EUR');
  });

  it('doesnt show defaults when specified', async () => {
    render(
      <AccountForm
        defaultValues={{
          name: 'Test account',
          parent: assetAccount,
          type: 'BANK',
          fk_commodity: eur,
        }}
        hideDefaults
        onSave={() => {}}
      />,
    );

    await waitFor(() => expect(screen.getByLabelText('Name')).toHaveValue('Test account'));

    const fieldsets = screen.getAllByRole('group');
    // Can't check with toBeVisible due tailwindcss not being understood by jest
    expect(fieldsets[2]).toHaveClass('hidden');
    expect(fieldsets[3]).toHaveClass('hidden');
    expect(fieldsets[5]).toHaveClass('hidden');
  });

  it('button is disabled when form not valid', async () => {
    const user = userEvent.setup();
    render(
      <AccountForm
        onSave={() => {}}
      />,
    );

    await user.type(screen.getByLabelText('Name'), 'ha');

    const button = await screen.findByText('add');
    await user.click(button);
    expect(button).toBeDisabled();
  });

  /**
   * INVESTMENT accounts can't have children
   */
  it('filters stock/mutual accounts as parents', async () => {
    const user = userEvent.setup();
    const commodity = await Commodity.create({
      mnemonic: 'TICKER',
      namespace: 'STOCK',
    }).save();

    const stockAccount = await Account.create({
      guid: 'stock_guid_1',
      name: 'Stock',
      type: 'INVESTMENT',
      fk_commodity: commodity,
      parent: assetAccount,
    }).save();

    stockAccount.path = 'Assets:Stock';

    jest.spyOn(apiHook, 'useAccounts').mockReturnValue({
      data: [
        root,
        assetAccount,
        expenseAccount,
        stockAccount,
      ],
    } as UseQueryResult<Account[]>);

    render(
      <AccountForm
        onSave={() => {}}
      />,
    );

    await user.click(screen.getByRole('combobox', { name: 'parentInput' }));
    screen.getByText('Assets');
    screen.getByText('Expenses');
    expect(screen.queryByText('Stock')).toBeNull();
    expect(screen.queryByText('Mutual')).toBeNull();
  });

  it('creates account with expected params and saves', async () => {
    const user = userEvent.setup();
    const mockSave = jest.fn();

    render(
      <AccountForm
        onSave={mockSave}
      />,
    );

    await user.type(screen.getByLabelText('Name'), 'TestAccount');

    await user.click(screen.getByRole('combobox', { name: 'parentInput' }));
    await user.click(screen.getByText('Assets'));

    await user.click(screen.getByRole('combobox', { name: 'typeInput' }));
    await user.click(screen.getByText('ASSET'));

    await user.click(screen.getByRole('combobox', { name: 'commodityInput' }));
    await user.click(screen.getByText('EUR'));

    expect(screen.getByText('add')).not.toBeDisabled();
    await user.click(screen.getByText('add'));

    const account = await Account.findOneByOrFail({ name: 'TestAccount' });
    expect(account).toEqual({
      guid: expect.any(String),
      name: 'TestAccount',
      type: 'ASSET',
      fk_commodity: eur,
      childrenIds: [],
      description: null,
      parentId: assetAccount.guid,
      path: 'Assets:TestAccount',
      placeholder: false,
      hidden: false,
    });
    expect(mockSave).toHaveBeenCalledTimes(1);
  });

  it('updates account', async () => {
    const user = userEvent.setup();
    const mockSave = jest.fn();

    const account = await Account.findOneOrFail({
      where: { name: 'Assets' },
      relations: { parent: true },
    });

    render(
      <AccountForm
        action="update"
        onSave={mockSave}
        defaultValues={{
          ...account,
        }}
      />,
    );

    await user.clear(screen.getByLabelText('Name'));
    await user.type(screen.getByLabelText('Name'), 'New name');

    expect(screen.getByText('update')).not.toBeDisabled();
    await user.click(screen.getByText('update'));
    const accounts = await Account.find();

    expect(accounts).toHaveLength(3);
    expect(accounts[1]).toMatchObject({
      guid: assetAccount.guid,
      name: 'New name',
      path: 'New name',
    });
  });

  it('deletes account', async () => {
    const user = userEvent.setup();
    const mockSave = jest.fn();

    const mockRouterReplace = jest.fn();
    jest.spyOn(navigation, 'useRouter').mockImplementation(() => ({
      replace: mockRouterReplace as AppRouterInstance['replace'],
    } as AppRouterInstance));

    const account = await Account.findOneOrFail({
      where: { name: 'Assets' },
      relations: { parent: true },
    });

    render(
      <AccountForm
        action="delete"
        onSave={mockSave}
        defaultValues={{
          ...account,
        }}
      />,
    );

    const deleteButton = await screen.findByText('delete');

    expect(deleteButton).not.toBeDisabled();
    await user.click(deleteButton);
    const accounts = await Account.find();

    expect(accounts).toHaveLength(2);
    expect(account.guid in accounts).toBe(false);
  });

  it('creates bank account with opening balance', async () => {
    const user = userEvent.setup();
    const mockSave = jest.fn();

    render(
      <AccountForm
        onSave={mockSave}
      />,
    );

    await user.type(screen.getByLabelText('Name'), 'TestAccount');

    await user.click(screen.getByRole('combobox', { name: 'parentInput' }));
    await user.click(screen.getByText('Assets'));

    await user.click(screen.getByRole('combobox', { name: 'typeInput' }));
    await user.click(screen.getByText('BANK'));

    await user.type(screen.getByLabelText('Opening balance'), '1000.10');
    await user.clear(screen.getByLabelText('When'));
    await user.type(screen.getByLabelText('When'), '2023-01-01');

    await user.click(screen.getByRole('combobox', { name: 'commodityInput' }));
    await user.click(screen.getByText('EUR'));

    expect(screen.getByText('add')).not.toBeDisabled();
    await user.click(screen.getByText('add'));

    const account = await Account.findOneByOrFail({ name: 'TestAccount' });
    expect(account).toEqual({
      guid: expect.any(String),
      name: 'TestAccount',
      type: 'BANK',
      fk_commodity: eur,
      childrenIds: [],
      description: null,
      parentId: assetAccount.guid,
      path: 'Assets:TestAccount',
      placeholder: false,
      hidden: false,
    });

    const txs = await Transaction.find();
    expect(txs).toHaveLength(1);
    expect(txs[0]).toMatchObject({
      guid: expect.any(String),
      description: 'Opening balance',
      fk_currency: {
        mnemonic: 'EUR',
      },
      date: DateTime.fromISO('2023-01-01'),
    });

    const equity = await Account.findOneByOrFail({ name: 'Opening balances - EUR' });
    const splits = await Split.find();
    expect(splits).toEqual([
      {
        accountId: account.guid,
        txId: expect.any(String),
        action: '',
        guid: expect.any(String),
        quantityDenom: 10,
        quantityNum: 10001,
        valueDenom: 10,
        valueNum: 10001,
      },
      {
        accountId: equity.guid,
        txId: expect.any(String),
        action: '',
        guid: expect.any(String),
        quantityDenom: 10,
        quantityNum: -10001,
        valueDenom: 10,
        valueNum: -10001,
      },
    ]);
  });

  it.each([
    ['EXPENSE', 'Expenses'],
    ['INCOME', 'Income'],
    ['ASSET', 'Assets'],
    ['BANK', 'Assets:Bank'],
  ])('filters selection for account type with %s parent', async (parentType, parentName) => {
    const incomeAccount = await Account.create({
      guid: 'income_guid_1',
      name: 'Income',
      type: 'INCOME',
      fk_commodity: eur,
      parent: root,
    }).save();
    incomeAccount.path = 'Income';

    const bankAccount = await Account.create({
      guid: 'bank_guid_1',
      name: 'Bank',
      type: 'BANK',
      fk_commodity: eur,
      parent: assetAccount,
    }).save();
    bankAccount.path = 'Assets:Bank';

    jest.spyOn(apiHook, 'useAccounts').mockReturnValue({
      data: [
        root,
        assetAccount,
        expenseAccount,
        incomeAccount,
        bankAccount,
      ],
    } as UseQueryResult<Account[]>);

    const user = userEvent.setup();
    render(
      <AccountForm
        onSave={() => {}}
      />,
    );

    await user.type(screen.getByLabelText('Name'), 'My account');

    await user.click(screen.getByRole('combobox', { name: 'parentInput' }));
    await user.click(screen.getByText(parentName));

    const typeInput = await screen.findByRole('combobox', { name: 'typeInput' });
    await user.click(typeInput);

    const allowedTypes = getAllowedSubAccounts(parentType);
    const ignoredTypes = allowedTypes.filter(type => !allowedTypes.includes(type));

    allowedTypes.forEach(type => screen.getByText(type));
    ignoredTypes.forEach(type => expect(screen.queryByText(type)).toBeNull());
  });
});
