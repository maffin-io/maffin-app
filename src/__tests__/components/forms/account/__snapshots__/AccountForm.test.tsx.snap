// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm renders as expected 1`] = `
<div>
  <form>
    <fieldset
      class="text-sm my-5"
    >
      <label
        class="inline-block mb-2"
        for="nameInput"
      >
        Name
      </label>
      <input
        class="block w-full m-0"
        id="nameInput"
        name="name"
        type="text"
      />
      <p
        class="invalid-feedback"
      />
    </fieldset>
    <fieldset
      class="text-sm my-5"
    >
      <label
        class="inline-block mb-2"
        for="parent"
      >
        Parent
      </label>
      <div
        class=" css-b62m3t-container"
        id="parentInput"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class="!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none css-13cymwt-control"
        >
          <svg
            class="absolute text-md left-[10px]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
            />
          </svg>
          <div
            class=" css-1fdsijx-ValueContainer"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-2-placeholder"
            >
              &lt;parent account&gt;
            </div>
            <div
              class="!text-inherit css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="parentInput"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class="hidden css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="parentInput"
          type="hidden"
          value=""
        />
      </div>
      <p
        class="invalid-feedback"
      />
    </fieldset>
    <fieldset
      class="text-sm my-5"
    >
      <label
        class="inline-block mb-2"
        for="type"
      >
        Account type
      </label>
      <div
        class=" css-3iigni-container"
        id="typeInput"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class="!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-700 !border-none css-16xfy0z-control"
        >
          <svg
            class="absolute text-md left-[10px]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 17h-1.559l-9.7-10.673A1 1 0 0 0 5.001 6H2v2h2.559l4.09 4.5-4.09 4.501H2v2h3.001a1 1 0 0 0 .74-.327L10 13.987l4.259 4.686a1 1 0 0 0 .74.327H17v3l5-4-5-4v3z"
            />
            <path
              d="M15.441 8H17v3l5-3.938L17 3v3h-2.001a1 1 0 0 0-.74.327l-3.368 3.707 1.48 1.346L15.441 8z"
            />
          </svg>
          <div
            class=" css-1fdsijx-ValueContainer"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-3-placeholder"
            >
              &lt;select parent first&gt;
            </div>
            <div
              class="!text-inherit css-1h01tm3-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-3-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="typeInput"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                disabled=""
                id="react-select-3-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class="hidden css-894a34-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <p
        class="invalid-feedback"
      />
    </fieldset>
    <fieldset
      class="text-sm my-5"
    >
      <label
        class="inline-block mb-2"
        for="commodity"
      >
        Commodity
      </label>
      <div
        class=" css-b62m3t-container"
        id="commodityInput"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-4-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class="!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none css-13cymwt-control"
        >
          <svg
            class="absolute text-md left-[10px]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 4H3a1 1 0 0 0-1 1v14a1 1 0 0 0 1 1h18a1 1 0 0 0 1-1V5a1 1 0 0 0-1-1zm-1 11a3 3 0 0 0-3 3H7a3 3 0 0 0-3-3V9a3 3 0 0 0 3-3h10a3 3 0 0 0 3 3v6z"
            />
            <path
              d="M12 8c-2.206 0-4 1.794-4 4s1.794 4 4 4 4-1.794 4-4-1.794-4-4-4zm0 6c-1.103 0-2-.897-2-2s.897-2 2-2 2 .897 2 2-.897 2-2 2z"
            />
          </svg>
          <div
            class=" css-1fdsijx-ValueContainer"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-4-placeholder"
            >
              &lt;commodity&gt;
            </div>
            <div
              class="!text-inherit css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-4-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="commodityInput"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-4-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class="hidden css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="commodityInput"
          type="hidden"
          value=""
        />
      </div>
      <p
        class="invalid-feedback"
      />
    </fieldset>
    <div
      class="flex w-full justify-center"
    >
      <button
        class="btn-primary"
        disabled=""
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
